File: leetcode/java/NextPermutation.java
public class NextPermutation {
    public void nextPermutation(int[] nums) {
        int pivot_idx = -1;
        for(int i = nums.length-1; i > 0; i++){
            if(nums[i-1]<nums[i]){
                pivot_idx = i;
                break;
            }
        } 
        //reverse the array 
        int l = pivot_idx, r = nums.length-1;
        while(l<r){
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++;
            r--;
        }
        if(pivot_idx!=-1){
            for(int i = pivot_idx; i < nums.length; i++){
                if(nums[i]>nums[pivot_idx-1])
            }
        }
        return ;
    }
}

File: leetcode/cpp/next-permutation.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
    void nextPermutation(vector<int> &nums){
        int pivot_idx = -1;
        for(int i = nums.size()-2; i >= 0; i--){
            if(nums[i] < nums[i+1]){
                pivot_idx=i;
                break;
            }
        }
        int first = pivot_idx+1, last = nums.size()-1;
        while(first<last){
            swap(nums[first++], nums[last--]);
        }
        if(pivot_idx!=-1){
            for(int i = pivot_idx+1;i < nums.size(); i++){
                if(nums[i] > nums[pivot_idx]){
                    swap(nums[i], nums[pivot_idx]);
                    return;
                }
            }
        }
        return;
    }
};